### PR TITLE
drivers: adc: add AD7799 IIO driver

### PR DESCRIPTION
**Pull request by amiclaus**
Add minimal implementation of the ad7799 IIO driver.

The functions implemented are tested, while the driver is subject for
further extension.

Signed-off-by: Antoniu Miclaus <antoniu.miclaus@analog.com>

### CODE REVIEW

**Comment by amiclaus** [comment]
V2:
- fix astyle

**Comment by amiclaus** [comment]
V3:
- include "iio_types.h" into the drivers makefile 
- remove unused variable `ch`

File: drivers/adc/ad7799/iio_ad7799.c [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+#include <string.h>
+#include "ad7799.h"
+#include "util.h"
+#include "error.h"
+
+/**
+ * @brief Read ADC Register.
+ * @param device - Device driver descriptor.
+ * @param buf - Input buffer.
+ * @param len - Length of the input buffer (not used in this case).
```
Comment (buha) [inline drivers/adc/ad7799/iio_ad7799.c:52]: looks like len is used in snprintf

File: drivers/adc/ad7799/iio_ad7799.c [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+ssize_t ad7799_iio_gain_write(void *device, char *buf, size_t len,
+			      const struct iio_ch_info *channel)
+{
+	struct ad7799_dev *dev = (struct ad7799_dev *)device;
+
+	uint32_t gain = srt_to_uint32(buf);
+
+	ad7799_set_gain(dev, (uint8_t)gain);
+
+	return len;
```
Comment (buha) [inline drivers/adc/ad7799/iio_ad7799.c:124]: looks like there should be some more error handling throughout the file.
what if the driver functions return failure ?

**Comment by amiclaus** [comment]
V4:
- fix header comments
- add return error code checks were possible.

File: drivers/adc/ad7799/iio_ad7799.h [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+		.shift = 0,
+		.is_big_endian = false
+	},
+	.attributes = ad7799_iio_channel_attributes,
+	.ch_out = false,
+};
+
+/** IIO ADC Channels */
+static struct iio_channel *ad7799_iio_channels[] = {
+	&ad7799_iio_channel_1,
```
[the remark this answers is not in the document]
Comment (amiclaus) [reply drivers/adc/ad7799/iio_ad7799.h:118]: Currently the branch is 2 commits behind master:
https://github.com/analogdevicesinc/no-OS/commit/cf332d83f05c20b340edd12659da49dabe07fb66
https://github.com/analogdevicesinc/no-OS/commit/0dd325aff0f2671e4bbcdc6bdbf956b7b2cbed82

[the remark this answers is not in the document]
Comment (amiclaus) [reply drivers/adc/ad7799/iio_ad7799.h:118]: And the commits are not related to iio stuff

**Comment by amiclaus** [comment]
V5:
- update driver to the latest IIO channel/attributes changes.

File: drivers/adc/ad7799/iio_ad7799.h [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	.sign = 'u',
+	.realbits = 12,
+	.storagebits = 24,
+	.shift = 0,
+	.is_big_endian = false
+};
+
+/** IIO ADC Channels */
+struct iio_channel ad7799_iio_channels[] = {
+	{
```
Comment (adrimbarean) [inline drivers/adc/ad7799/iio_ad7799.h:79]: nitpick: if you make a define for channel declaration you can avoid a lot of code duplication.

Comment (amiclaus) [reply drivers/adc/ad7799/iio_ad7799.h:79]: Sure, I can do :)

File: drivers/Makefile [hunk 5]
```diff
@@ -16,7 +16,8 @@ INCLUDES = -I../include/ \
 	 -I./axi_core/axi_dmac \
 	 -I./axi_core/clk_axi_clkgen \
 	 -I./platform/xilinx \
-	 -I../projects/adrv9009/src/devices/adi_hal
+	 -I../projects/adrv9009/src/devices/adi_hal \
+	 -I../libraries/iio
```
Comment (mchindri) [inline drivers/Makefile:20]: Is this change related to this commit?

File: drivers/adc/ad7799/iio_ad7799.h [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+*******************************************************************************/
+
+#ifndef IIO_AD7799_H
+#define IIO_AD7799_H
+
+#include "iio_types.h"
+
+/** AD7799 IIO register read */
+ssize_t ad7799_iio_reg_read(void *device, char *buf, size_t len,
+			    const struct iio_ch_info *channel, intptr_t priv);
```
Comment (mchindri) [inline drivers/adc/ad7799/iio_ad7799.h:47]: If you use debug_reg_read, this functions if not need anymore. Neither ad7799_iio_debug_attributes

File: drivers/Makefile (hunk 5 above)
Comment (amiclaus) [reply drivers/Makefile:20]: If I move the change to different commit, the build for this particular commit will fail.

Comment (mchindri) [reply drivers/Makefile:20]: Sholdn't then the iio file be added in /noos/iio ?

Comment (amiclaus) [reply drivers/Makefile:20]: Based on a previous meeting discussion, we decided to add the IIO driver in the driver's folder.

Comment (mchindri) [reply drivers/Makefile:20]: Ok, I missed the meeting :)

Comment (mchindri) [reply drivers/Makefile:20]: Anyway, you can add a pr for this change, in #863 will be needed too.

**Comment by amiclaus** [comment]
V6:
- remove makefile changes
- remove debug attributes and reg_read function
- add define for channel declaration

File: drivers/adc/ad7799/iio_ad7799.h (hunk 4 above)
Comment (mchindri) [reply drivers/adc/ad7799/iio_ad7799.h:79]: I think Andrei was meaning something like this:
image

File: drivers/adc/ad7799/iio_ad7799.c [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+{
+	struct ad7799_dev *dev = (struct ad7799_dev *)device;
+	uint32_t data;
+	int32_t ret;
+
+	uint32_t addr = srt_to_uint32(buf);
+
+	ret = ad7799_read(dev, addr, &data);
+	if (ret != SUCCESS)
+		return FAILURE;
```
Comment (mchindri) [inline drivers/adc/ad7799/iio_ad7799.c]: return ret;
This why error code can be viewed in an upper level

File: drivers/adc/ad7799/iio_ad7799.h [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#ifndef IIO_AD7799_H
+#define IIO_AD7799_H
+
+#include "iio_types.h"
+
+/** AD7799 IIO register read */
+ssize_t ad7799_iio_reg_read(void *device, char *buf, size_t len,
```
[the remark this answers is not in the document]
Comment (mchindri) [reply drivers/adc/ad7799/iio_ad7799.h]: You didn't remove it

File: drivers/adc/ad7799/iio_ad7799.h (hunk 4 above)
Comment (amiclaus) [reply drivers/adc/ad7799/iio_ad7799.h]: The old file was overwritten locally when I applied the astyle.
should be fine now.

**Comment by amiclaus** [comment]
V7:
- return FAILURE -> return ret

**Review by mchindri — APPROVED** [review]

**Comment by dbogdan** [comment]
There is still an astyle issue.

**Review by dbogdan — APPROVED** [review]

**Comment by amiclaus** [comment]
V8:
- more astyle fixes
